File: apps/fintr-be/app/utils/utils/installment_plan.rb
# frozen_string_literal: true

# Installment split and revision math. Canonical behavior:
# docs/installment_plans.md
module Utils
  class InstallmentPlan
    ANCHORS = %w[total monthly explicit].freeze

    class << self
      include Dry::Monads[:result]

      def occurrence_dates(parent_date:, period:, from_date: nil)
        from = from_date || parent_date

        (0...period).map { |i| parent_date + i.months }.select { |d| d >= from }
      end

      def remaining_occurrence_dates(
        parent_date:,
        period:,
        effective_date:,
        calculated_dates: []
      )
        occurrence_dates(
          parent_date:,
          period:,
          from_date: effective_date,
        )
      end

      def frozen_occurrence_dates(
        parent_date:,
        period:,
        effective_date:,
        calculated_dates: []
      )
        occurrence_dates(
          parent_date:,
          period:,
        ).reject { |date| date >= effective_date }
      end

      def paid_so_far_cents(series_transactions:)
        series_transactions.where(balance_state: :calculated).sum(:amount_cents)
      end

      def resolve_installment_total_cents(root:, series_transactions: nil)
        return root.installment_total_cents if root.installment_total_cents.present?

        series = series_transactions || root.series_records
        summed_cents = series.sum(:amount_cents)
        return summed_cents if summed_cents.positive?

        if root.installment_period.present? && root.installment_period.positive?
          root.amount_cents * root.installment_period
        end
      end

      # Plan total lives on the series root. Child rows may carry a stale copy.
      def series_installment_total_cents(transaction:)
        root = transaction.respond_to?(:root_parent) ? transaction.root_parent : transaction
        root&.installment_total_cents.presence || transaction.installment_total_cents
      end

      def series_installment_total_amount(transaction:)
        cents = series_installment_total_cents(transaction:)
        return nil if cents.blank?

        Money.new(cents, transaction.amount_currency).amount
      end

      def to_ledger_amount(
        amount:,
        from_currency:,
        ledger_currency:,
        exchange_rate: nil
      )
        cents = to_ledger_cents(
          amount:,
          from_currency:,
          ledger_currency:,
          exchange_rate:,
        )
        return nil if cents.blank?

        Money.new(cents, ledger_currency).amount
      end

      def to_ledger_cents(
        amount:,
        from_currency:,
        ledger_currency:,
        exchange_rate: nil
      )
        return nil if amount.blank?

        from = from_currency.to_s.strip.upcase
        ledger = ledger_currency.to_s.strip.upcase
        rate =
          if exchange_rate.present?
            BigDecimal(exchange_rate.to_s)
          end
        ledger_amount =
          if from.present? &&
             ledger.present? &&
             from != ledger &&
             rate.present? &&
             rate.positive?
            (BigDecimal(amount.to_s) * rate).round(2)
          else
            BigDecimal(amount.to_s)
          end

        Money.from_amount(ledger_amount, ledger_currency).cents
      end

      def commitment_total_cents(
        parent_date:,
        period:,
        default_per_payment_cents:,
        occurrence_cents_by_date:
      )
        occurrence_dates(
          parent_date:,
          period:,
        ).sum do |date|
          occurrence_cents_by_date[date] ||
            occurrence_cents_by_date[date.iso8601] ||
            default_per_payment_cents
        end
      end

      def apply_this_only_amount_delta_cents(
        stored_total_cents:,
        period:,
        previous_amount_cents:,
        next_amount_cents:
      )
        delta = next_amount_cents.to_i - previous_amount_cents.to_i
        base =
          if stored_total_cents.present? && stored_total_cents.to_i.positive?
            stored_total_cents.to_i
          else
            previous_amount_cents.to_i * period.to_i
          end

        base + delta
      end

      def compute_revision(
        installment_total_cents:,
        currency:,
        new_period:,
        parent_date:,
        effective_date:,
        anchor:,
        paid_so_far_cents:,
        new_monthly_cents: nil,
        calculated_dates: [],
        prior_per_payment_cents: 0,
        occurrence_cents_by_date: {}
      )
        return Failure(anchor: "is invalid") unless ANCHORS.include?(anchor)

        remaining_dates = remaining_occurrence_dates(
          parent_date:,
          period: new_period,
          effective_date:,
          calculated_dates:,
        )
        remaining_count = remaining_dates.size
        if remaining_count.zero?
          return Failure(installment_period: "must leave at least one remaining payment")
        end

        frozen_dates = frozen_occurrence_dates(
          parent_date:,
          period: new_period,
          effective_date:,
        )
        occurrence_lookup = normalize_occurrence_cents_by_date(occurrence_cents_by_date)
        locked_cents = frozen_dates.sum do |date|
          occurrence_lookup[date] ||
            occurrence_lookup[date.iso8601] ||
            prior_per_payment_cents.to_i
        end
        paid_money = Money.new(locked_cents, currency)

        case anchor
        when "total"
          total_money = Money.new(installment_total_cents, currency)
          remaining_total = total_money - paid_money
          per_payment = divide_money_evenly(
            remaining_cents: remaining_total.cents,
            count: remaining_count,
          )
          new_total_cents = installment_total_cents
        when "monthly"
          monthly_cents = new_monthly_cents
          return Failure(amount: "is required for keep monthly revision") if monthly_cents.blank?

          new_total_cents = locked_cents + (monthly_cents * remaining_count)
          per_payment = Money.new(monthly_cents, currency).amount
        when "explicit"
          total_money = Money.new(installment_total_cents, currency)
          remaining_total = total_money - paid_money
          per_payment = divide_money_evenly(
            remaining_cents: remaining_total.cents,
            count: remaining_count,
          )
          new_total_cents = installment_total_cents
        end

        Success(
          installment_total_cents: new_total_cents,
          per_payment: per_payment,
          remaining_count: remaining_count,
          paid_so_far_cents: paid_so_far_cents,
        )
      end

      private

      def divide_money_evenly(remaining_cents:, count:)
        return BigDecimal("0") if count.zero?

        (remaining_cents.to_d / count / 100).round(2, BigDecimal::ROUND_HALF_UP)
      end

      def normalize_occurrence_cents_by_date(occurrence_cents_by_date)
        return {} if occurrence_cents_by_date.blank?

        occurrence_cents_by_date.each_with_object({}) do |(date, cents), lookup|
          next if cents.blank?

          key =
            case date
            when Date
              date
            when String
              Date.iso8601(date)
            else
              date.respond_to?(:to_date) ? date.to_date : date
            end
          lookup[key] = cents.to_i
          lookup[key.iso8601] = cents.to_i if key.respond_to?(:iso8601)
        end
      end
    end
  end
end

File: packages/fintr-domain/src/installment-plan.ts
/** Shared installment split / revision math. Canonical behavior: docs/installment_plans.md */

export const INSTALLMENT_REVISION_ANCHORS = ["total", "monthly", "explicit"] as const;

export type InstallmentRevisionAnchor = (typeof INSTALLMENT_REVISION_ANCHORS)[number];

export type InstallmentPlanRevisionInput = {
  installmentTotalCents: number;
  currency: string;
  newPeriod: number;
  parentDate: string;
  effectiveDate: string;
  anchor: InstallmentRevisionAnchor;
  paidSoFarCents: number;
  newMonthlyCents?: number | null;
  calculatedDates?: string[];
  /** Fallback per frozen month when that date is missing from occurrenceCentsByDate. */
  priorPerPaymentCents?: number | null;
  /**
   * Actual amounts by occurrence date. Frozen months (before effectiveDate) use
   * these so a this-only bump on the revision anchor is released into the
   * remaining pool instead of locking earlier months at the bumped rate.
   */
  occurrenceCentsByDate?: Record<string, number>;
};

export type InstallmentPlanRevisionResult = {
  installmentTotalCents: number;
  perPayment: number;
  remainingCount: number;
  paidSoFarCents: number;
};

const roundHalfUp = (value: number, decimals = 2): number => {
  const factor = 10 ** decimals;
  return Math.round((value + Number.EPSILON) * factor) / factor;
};

const parseDate = (value: string): Date => {
  const dateOnly = value.slice(0, 10);
  const [year, month, day] = dateOnly.split("-").map(Number);
  return new Date(Date.UTC(year, month - 1, day));
};

const formatDate = (date: Date): string => {
  const year = date.getUTCFullYear();
  const month = `${date.getUTCMonth() + 1}`.padStart(2, "0");
  const day = `${date.getUTCDate()}`.padStart(2, "0");
  return `${year}-${month}-${day}`;
};

const addMonthsUtc = (date: Date, months: number): Date => {
  const next = new Date(date.getTime());
  next.setUTCMonth(next.getUTCMonth() + months);
  return next;
};

export const installmentOccurrenceDates = ({
  parentDate,
  period,
  fromDate,
}: {
  parentDate: string;
  period: number;
  fromDate?: string;
}): string[] => {
  const parent = parseDate(parentDate);
  const from = fromDate ? parseDate(fromDate) : parent;
  const dates: string[] = [];

  for (let index = 0; index < period; index += 1) {
    const occurrence = addMonthsUtc(parent, index);
    if (occurrence.getTime() >= from.getTime()) {
      dates.push(formatDate(occurrence));
    }
  }

  return dates;
};

export const installmentRemainingOccurrenceDates = ({
  parentDate,
  period,
  effectiveDate,
}: {
  parentDate: string;
  period: number;
  effectiveDate: string;
  calculatedDates?: string[];
}): string[] =>
  installmentOccurrenceDates({
    parentDate,
    period,
    fromDate: effectiveDate,
  });

export const installmentFrozenOccurrenceDates = ({
  parentDate,
  period,
  effectiveDate,
}: {
  parentDate: string;
  period: number;
  effectiveDate: string;
  calculatedDates?: string[];
}): string[] => {
  const effectiveTime = parseDate(effectiveDate).getTime();

  return installmentOccurrenceDates({
    parentDate,
    period,
  }).filter((date) => parseDate(date).getTime() < effectiveTime);
};

export const computeInstallmentPlanRevision = (
  input: InstallmentPlanRevisionInput,
): InstallmentPlanRevisionResult => {
  const remainingDates = installmentRemainingOccurrenceDates({
    parentDate: input.parentDate,
    period: input.newPeriod,
    effectiveDate: input.effectiveDate,
  });
  const remainingCount = remainingDates.length;

  if (remainingCount === 0) {
    throw new Error("Installment term must leave at least one remaining payment");
  }

  const frozenDates = installmentFrozenOccurrenceDates({
    parentDate: input.parentDate,
    period: input.newPeriod,
    effectiveDate: input.effectiveDate,
  });
  const priorPerPaymentCents = input.priorPerPaymentCents ?? 0;
  const occurrenceCentsByDate = input.occurrenceCentsByDate ?? {};
  const lockedCents = frozenDates.reduce(
    (sum, date) =>
      sum + (occurrenceCentsByDate[date] ?? priorPerPaymentCents),
    0,
  );
  const paidSoFar = lockedCents / 100;

  if (input.anchor === "monthly") {
    const monthlyCents = input.newMonthlyCents;
    if (monthlyCents == null) {
      throw new Error("Monthly amount is required for keep monthly revision");
    }

    const perPayment = monthlyCents / 100;
    const installmentTotalCents = lockedCents + monthlyCents * remainingCount;

    return {
      installmentTotalCents,
      perPayment,
      remainingCount,
      paidSoFarCents: input.paidSoFarCents,
    };
  }

  const total = input.installmentTotalCents / 100;
  const remainingTotal = total - paidSoFar;
  const perPayment = roundHalfUp(remainingTotal / remainingCount);

  return {
    installmentTotalCents: input.installmentTotalCents,
    perPayment,
    remainingCount,
    paidSoFarCents: input.paidSoFarCents,
  };
};

export const toLedgerCents = ({
  amount,
  fromCurrency,
  ledgerCurrency,
  exchangeRate,
}: {
  amount: number;
  fromCurrency?: string | null;
  ledgerCurrency?: string | null;
  exchangeRate?: number | null;
}): number => {
  if (!Number.isFinite(amount)) {
    return 0;
  }

  const from = fromCurrency?.trim().toUpperCase() ?? "";
  const ledger = ledgerCurrency?.trim().toUpperCase() ?? "";
  const rate = Number(exchangeRate);

  if (
    from
    && ledger
    && from !== ledger
    && Number.isFinite(rate)
    && rate > 0
  ) {
    return Math.round(amount * rate * 100);
  }

  return Math.round(amount * 100);
};

export const resolveInstallmentTotalCents = ({
  installmentTotalCents,
  perPaymentCents,
  period,
  seriesAmountCents,
}: {
  installmentTotalCents?: number | null;
  perPaymentCents: number;
  period: number;
  seriesAmountCents?: number | null;
}): number => {
  if (installmentTotalCents != null && installmentTotalCents > 0) {
    return installmentTotalCents;
  }

  if (seriesAmountCents != null && seriesAmountCents > 0) {
    return seriesAmountCents;
  }

  return perPaymentCents * period;
};

export const computeInstallmentCommitmentTotalCents = ({
  parentDate,
  period,
  defaultPerPaymentCents,
  occurrenceCentsByDate,
}: {
  parentDate: string;
  period: number;
  defaultPerPaymentCents: number;
  occurrenceCentsByDate: Record<string, number>;
}): number => {
  if (period <= 0) {
    return 0;
  }

  return installmentOccurrenceDates({
    parentDate,
    period,
  }).reduce(
    (sum, date) => sum + (occurrenceCentsByDate[date] ?? defaultPerPaymentCents),
    0,
  );
};

export const applyInstallmentThisOnlyTotalDeltaCents = ({
  storedTotalCents,
  period,
  previousAmountCents,
  nextAmountCents,
}: {
  storedTotalCents?: number | null;
  period: number;
  previousAmountCents: number;
  nextAmountCents: number;
}): number => {
  const delta = nextAmountCents - previousAmountCents;
  const base =
    storedTotalCents != null && storedTotalCents > 0
      ? storedTotalCents
      : previousAmountCents * period;

  return base + delta;
};
